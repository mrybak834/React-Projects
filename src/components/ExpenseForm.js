import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

export default class ExpenseForm extends React.Component {

    constructor(props){
        super(props);

        this.state = {
            description: '',
            note: '',
            ...props.expense,
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            timestamp: props.expense ? moment(props.expense.timestamp) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;

        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;

        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        // Accept no amount or a valid value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onCalendarDateChange = (timestamp) => {
        timestamp && this.setState(() => ({ timestamp }));
    };

    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();

        // Handle input errors
        this.setState(() => ({ error: '' }));
        if (!this.state.description.trim()){
            return this.setState(() => ({ error: 'Please enter a description!'}));
        }
        if (!this.state.amount){
            return this.setState(() => ({ error: 'Please enter an amount!'}));
        }

        this.props.onSubmit({
            description: this.state.description,
            note: this.state.note,
            amount: parseFloat(this.state.amount, 10) * 100,
            timestamp: this.state.timestamp.valueOf()
        });
    };

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input 
                        id="inputDescription"
                        type="text" 
                        placeholder="Description" 
                        autoFocus 
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        id="inputAmount"
                        type="text" 
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.timestamp}
                        onDateChange={this.onCalendarDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (Optional)" 
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button >Add Expense</button>
                </form>
                {this.state.error && <p>{this.state.error}</p>}
            </div>
        )
    }
}