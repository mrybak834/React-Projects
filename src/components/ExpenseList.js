import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../redux/selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                props.expenses.map((expense, index) => 
                    <ExpenseListItem 
                        key={index}
                        {...expense}/>
                )
            )
        }
    </div>
);


const mapStateToProps = (state) => ({
        expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);