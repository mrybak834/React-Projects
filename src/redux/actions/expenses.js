import uuid from 'uuid';
import database from '../../firebase/firebase';
/**
 * Normal redux actions:
 * Component calls the action generator (these functions)
 * Actinn generator returns the generated object
 * Component dispatches the object
 * Redux store changes/provides info
 * 
 * Asynchronous actions (for interacting with DB):
 * Component calls action generator
 * Action generator returns a function
 * Component dispatches a function (done via redux-thunk)
 * function runs, interacts with db, can do whatever else it wants,
 * including then generating another action to dispatch to the traditional redux store
 */

export const addExpense = (expense) => (
    {
        type: 'ADD_EXPENSE',
        expense
    }
);

/**
 * Explanation of syntax:
 * (expenseData = {})          Takes the expense passed in. If none, set to {}
 * 
 * const {...} = expenseData   Creates the variables with the values.
 *                             Any variable that expenseData contains, overrides the LHS variables.
 * 
 * Equivalent to: 
 * const fun = 
 * ( 
 *     {
 *         variables declared
 *     } = {} 
 * ) => { variables available here }
 * 
 */
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            timestamp = 0
        } = expenseData;
        const expense = { description, note, amount, timestamp };

        // This promise is resolved before returning
        // Since it dispatches to the store, and we are hooked up to a fake store in our test cases
        // We can access the result of the dispatch call in our tests, as a promise
        // Since we return it, this result can be (and is) promise chained for our tests
        return database.ref('expenses')
            .push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            })
            .catch((e) => console.log('Adding expense failed: ', e));
    };
}

export const removeExpense = (id = undefined) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
);

export const editExpense = (id = undefined, updates = {}) => (
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
);