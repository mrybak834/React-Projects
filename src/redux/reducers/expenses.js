const expensesReducerDefaultState = [];

export default ( state = expensesReducerDefaultState, action ) => {
    switch (action.type){
        case 'ADD_EXPENSE':{
            if (!action.expense){
                return state;
            }
            return state.concat(action.expense);
            // return [...state, action.expense];   // Slower than concat
        }
        case 'REMOVE_EXPENSE':{
            // No id defined, don't remove anything
            if (!action.id){
                return state;
            }

            return state.filter(({id}) => id !== action.id);
        }
        case 'EDIT_EXPENSE':{
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        // Overrides any of the expense properties above
                        ...action.updates
                    }
                }

                return expense;
            });
        }
        default:{
            return state;
        }
    }
};