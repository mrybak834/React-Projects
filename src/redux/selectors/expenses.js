import moment from 'moment';


export default (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter( (expense) => {
        // // Filter text
        // if ( text && ((expense.description !== text) && (expense.note !== text)) ){
        //     return false;
        // }
        // // Filter Dates
        // if ( (startDate && (expense.timestamp < startDate)) || (endDate && (expense.timestamp > endDate))){
        //     return false;
        // }

        const timestampMoment = moment(expense.timestamp);

        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = startDate ? startDate.isSameOrBefore(timestampMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(timestampMoment, 'day') : true;
        return textMatch && startDateMatch && endDateMatch;
    }).sort( (a,b) => {
        if (sortBy === 'date'){
            // -1 = a comes first
            return a.timestamp > b.timestamp ? -1 : 1;
        }
        // amount
        else {
            return a.amount > b.amount ? -1 : 1;
        }
    });
};