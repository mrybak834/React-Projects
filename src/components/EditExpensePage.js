import React from 'react';
import ExpenseForm from '../components/ExpenseForm';
import { connect } from 'react-redux';
import { startEditExpense } from '../redux/actions/expenses';
import { startRemoveExpense } from '../redux/actions/expenses';

export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);

        this.props.history.push('/');
    };

    onClick = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    };

    render() {
        return (

            this.props.expense ? (
                <div>
                    <div className="page-header">
                        <div className="content-container">
                            <h1 className="page-header__title">Edit Expense: {this.props.expense.description}</h1>
                        </div>
                    </div>
                    <div className="content-container">
                        <ExpenseForm
                            expense={this.props.expense}
                            onSubmit={this.onSubmit}
                        />
                        <button
                            className="button button--secondary"
                            onClick={this.onClick}
                        >Remove</button>
                    </div>
                </div>
            ) : (
                    <div className="page-header">
                        <div className="content-container">
                            <h1 className="page-header__title">Invalid Expense</h1>
                        </div>
                    </div>
                )
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);