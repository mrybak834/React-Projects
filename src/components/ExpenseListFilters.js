import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, setSortByFilter, setStartDateFilter, setEndDateFilter } from '../redux/actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    
    state = {
        calendarFocused: null
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
        this.props.setSortByFilter(e.target.value);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDateFilter(startDate);
        this.props.setEndDateFilter(endDate);
    };

    render () {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
                
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSelectChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="start_date"
                    endDate={this.props.filters.endDate}
                    endDateId="end_date"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={2}
                    isOutsideRange={() => false}
                />
        
            </div>
        );
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: value => dispatch(setTextFilter(value)),
    setSortByFilter: value => dispatch(setSortByFilter(value)),
    setStartDateFilter: startDate => dispatch(setStartDateFilter(startDate)),
    setEndDateFilter: endDate => dispatch(setEndDateFilter(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);