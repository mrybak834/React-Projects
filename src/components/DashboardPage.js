import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';


const DashboardPage = () => (
    <div>
        <p>Dashboard Page</p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;