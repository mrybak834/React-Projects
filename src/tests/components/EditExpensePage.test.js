import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';


let history, editExpense, startRemoveExpense, expense, wrapper;

beforeEach(() => {
    history = {
        push: jest.fn()
    };
    editExpense = jest.fn();
    startRemoveExpense = jest.fn();
    expense = expenses[0];
    wrapper = shallow(<EditExpensePage expense={expense} history={history} editExpense={editExpense} startRemoveExpense={startRemoveExpense}/>);
});

test('Render Page', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Expense Form Submitted', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense);

    expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense);
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('Remove Button Clicked', () => {
    wrapper.find('button').prop('onClick')();

    expect(startRemoveExpense).toHaveBeenLastCalledWith(expense.id);
    expect(history.push).toHaveBeenLastCalledWith('/');
});