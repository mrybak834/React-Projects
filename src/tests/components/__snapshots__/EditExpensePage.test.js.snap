// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Page 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense: 
        test
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 1,
          "description": "test",
          "id": "1",
          "note": "test",
          "timestamp": 864000000,
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
